flasc/analysis/expected_power_analysis_utilities.py

```diff
@@ -0,0 +1,484 @@
+) -> pl.DataFrame:
+    """Fill covariance terms with the product of the square root of the variances.
+
+    Fill the null (or all) values in the covariance matrix with the product
+    of the square root of the variances of the corresponding test columns.
+
+    Leave the number of points as is (the number of shared points between the two test columns).
+
+    Args:
+        df_cov (pl.DataFrame): A polars dataframe with the covariance matrix
+        test_cols (List[str]): A list of column names to calculate the covariance of
+        fill_all (bool): If True, fill all values of cov, regardless of whether or not missing/Null
+
+    Returns:
+        pl.DataFrame: A polars dataframe with the null values filled according to the strategy.
+    """
+    n_test_cols = len(test_cols)
+
+    # Loop over all combinations of test columns
+    for t1_idx in range(n_test_cols):
+        for t2_idx in range(n_test_cols):
+            if t1_idx == t2_idx:
+                continue
+
+            t1 = test_cols[t1_idx]
+            t2 = test_cols[t2_idx]
+
+            # Get the cov_col, and the variance columns number of points for each turbine
+            cov_col = f"cov_{t1}_{t2}"
+            n_col = f"count_{t1}_{t2}"
+            var_1_col = f"cov_{t1}_{t1}"
+            var_2_col = f"cov_{t2}_{t2}"
+            n_1_col = f"count_{t1}_{t1}"
+            n_2_col = f"count_{t2}_{t2}"
+
+            # For the rows where cov_col is null, fill the cov_col with the product of the square
+            # root of the variances of the two test columns and the n_col with the minimum of the
+            # number of points for the two test columns
+            df_cov = df_cov.with_columns(null_map=pl.col(cov_col).is_null())
+
+            # If fill_all is true, set null_map True for all rows
+            if fill_all:
+                df_cov = df_cov.with_columns(pl.lit(True).alias("null_map"))
+
+            df_cov = df_cov.with_columns(
+                pl.when(pl.col("null_map"))
+                .then((pl.col(var_1_col).sqrt() * pl.col(var_2_col).sqrt()))
+                .otherwise(pl.col(cov_col))
+                .alias(cov_col)
+            )
+
+            # Num points as the minimum between the values of the two variances
+            df_cov = df_cov.with_columns(
+                pl.when(pl.col("null_map"))
+                .then(pl.min_horizontal(n_1_col, n_2_col))
+                .otherwise(pl.col(n_col))
+                .alias(n_col)
+            )
+
+            # For any rows where n_col is < 2 set n_col to null
+            df_cov = df_cov.with_columns(
+                pl.when(pl.col(n_col) < 2).then(None).otherwise(pl.col(n_col)).alias(n_col)
+            )
+
+            # Wherever n_col is null, set cov_col to null
+            df_cov = df_cov.with_columns(
+                pl.when(pl.col(n_col).is_null())
+                .then(None)
+                .otherwise(pl.col(cov_col))
+                .alias(cov_col)
+            )
```

Comment:
If n_col is Null, then won't cov_col already be null?

Reply:
I wasn't positive, in the tests I made I can make scenarios where n_col is Null but not cov_col (for that reason if you comment out the above block test_fill_cov_with_var_fill_all will fail), but maybe that is not a realistic condition to test for

Reply:
Ok removed